File: ruby/secret-handshake/secret_handshake.rb
class SecretHandshake
  attr_accessor :code

  def initialize(code)
    @code = code.to_i
  end

  def commands
    return [] unless valid_code?
    out = []
    out << 'wink' if set?(WINK)
    out << 'double blink' if set?(BLINK)
    out << 'close your eyes' if set?(CLOSE)
    out << 'jump' if set?(JUMP)
    out.reverse! if set?(REVERSE)
    out
  end

  private

  def set?(flag)
    @code & flag > 0
  end

  def valid_code?
    @code >= 0 && code <= 31
  end

  # Flags
  WINK = 1
  BLINK = 2
  CLOSE = 4
  JUMP = 8
  REVERSE = 16
end

File: scala/secret-handshake/src/main/scala/SecretHandshake.scala
import scala.util.Try

object SecretHandshake {
  val components = Map(
    0x1 -> "wink",
    0x2 -> "double blink",
    0x4 -> "close your eyes",
    0x8 -> "jump"
  )

  val reverseFlag = 0x10

  def handshake(code: Int): Seq[String] = {
    val actions = components.filterKeys(isSet(code, _)).values.toList
    if (isSet(code, reverseFlag))
      actions.reverse
    else
      actions
  }

  def handshake(code: String): Seq[String] =
    Try(handshake(Integer.parseInt(code, 2))).getOrElse(List())

  private def isSet(code: Int, flag: Int) = (code & flag) == flag
}
